[AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

ritter-x2a opened the pull request:
In the SelectionDAG lowering of the memcpy intrinsic, this optimization introduces additional chains between fixed-size groups of loads and the corresponding stores. While initially introduced to ensure that wider load/store-pair instructions are generated on AArch64, this optimization also improves code generation for AMDGPU: Ganged loads are scheduled into a clause; stores only await completion of their corresponding load.

The chosen value of 32 performed good in microbenchmarks, values of 8, 16, or 64 would perform similarly.
The testcase updates are autogenerated by utils/update_llc_test_checks.py.

See also:
 - PR introducing this optimization: https://reviews.llvm.org/D46477

Part of SWDEV-455845.

ritter-x2a committed:
[AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

In the SelectionDAG lowering of the memcpy intrinsic, this optimization
introduces additional chains between fixed-size groups of loads and the
corresponding stores. While initially introduced to ensure that wider
load/store-pair instructions are generated on AArch64, this optimization also
improves code generation for AMDGPU: Ganged loads are scheduled into a clause;
stores only await completion of their corresponding load.

The chosen value of 32 performed good in microbenchmarks, values of 8, 16, or
64 would perform similarly.
The testcase updates are autogenerated by utils/update_llc_test_checks.py.

See also:
 - PR introducing this optimization: https://reviews.llvm.org/D46477

Part of SWDEV-455845.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -67,6 +67,9 @@ AMDGPUTargetLowering::AMDGPUTargetLowering(const TargetMachine &TM,
   MaxStoresPerMemcpy = MaxStoresPerMemcpyOptSize = ~0U;
   MaxStoresPerMemmove = MaxStoresPerMemmoveOptSize = ~0U;
 
+  // Enable ganging up loads and stores in the memcpy DAG lowering.
```
Why does this need to be a target option? Why isn't the default just some number?

ritter-x2a replied:
The current default is 0, disabling this transformation. The original author only enabled it for AArch64 and kept the previous behavior for the other targets.

In general, I don't think that it is obvious that this "optimization" (in a loose sense) improves performance for every target; it effectively limits the later stages of codegen in its choices.
We would need to benchmark the other targets to be sure that we don't decrease their performance if we want it to be enabled by default.

arsenm replied:
I mean it's a problem when things get added for only a single target out of fear of regressing anything else. Also I don't see what target property would correspond to a value here. It's in terms of store count, not even bytes or load width or anything. I'm saying this is a bad API

arsenm replied:
We'd probably want different numbers based on the address spaces involved

ritter-x2a replied:
> I mean it's a problem when things get added for only a single target out of fear of regressing anything else. Also I don't see what target property would correspond to a value here. It's in terms of store count, not even bytes or load width or anything. I'm saying this is a bad API

Another argument for that: The limit as implemented is actually not concerning only stores, but loads and stores. So 32 here means that packets of 16 loads and 16 stores are ganged up.
I take it that you suggest that we improve the API before using it?

I think the corresponding hardware property for AMDGPU would be something like the number of memory accesses that can be started before the first one finishes.
In AArch64, it is the number of memory operations that can be merged into one (i.e. 2 for ldp and stp).
For architectures with traditional vector operations, it might be the preferred number of vector lanes.

> We'd probably want different numbers based on the address spaces involved

I'll do some benchmarks for other address spaces to find out, then.

arsenm replied:
Turning it on first is fine. We should also expand the test coverage for different sizes, alignments and address space combinations. I'm surprised we don't have more test updates from this as-is

ritter-x2a replied:
@arsenm I ran some benchmarks for the other address spaces; enabling this optimization either improved or did not change the performance compared to the original memcpy implementation in all cases.
Which exact parameter value among {8, 16, 32, 64} we choose does not make much of a difference with different address spaces either; I now changed the parameter value in the PR to 16 instead of the previous 32 since that performed a bit better when copying from the generic to the global address space.

At least in my benchmarks on gfx1030, different parameter values for different address spaces seem unnecessary. 

Regarding test coverage, I can easily generate llc tests for various parameter combinations; the resulting test can however grow quite large. For combinations of the following parameters, a test for memmove and memcpy with auto-generated llc check lines reaches ~60k lines (ca. 1.5x the length of the largest AMDGPU codegen test so far):
```
  dst_address_spaces = [0, 1, 3, 5]
  src_address_spaces = [0, 1, 3, 4, 5]
  dst_alignments = [1, 2, 4]
  src_alignments = [1, 2, 4]
  sizes = [16, 31, 32]
```
Would that be a useful addition to our testing?

arsenm replied:
Testing 8 and 16 alignments will at least show some LDS differences. We can probably get away without the addrspace 4 cases, unless they are specially crafted to use uniform source pointers

ritter-x2a replied:
I just added a generated test (with alignments 8 and 16 instead of 4, leading to 97k lines) to the PR, so that you can see what it looks like. Running it alone costs ca. 5 seconds on my workstation.

ritter-x2a replied:
Looking more into the code, I noticed that the DAG lowering of memcpy only takes the smaller one when different alignments are specified for source and destination. So, there is no point in testing combinations of different alignments, which leaves us with ~20k lines of test cases (the now updated state in this PR).

Do you have objections against merging the PR like this, @arsenm ?

ritter-x2a committed:
amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

[AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

In the SelectionDAG lowering of the memcpy intrinsic, this optimization
introduces additional chains between fixed-size groups of loads and the
corresponding stores. While initially introduced to ensure that wider
load/store-pair instructions are generated on AArch64, this optimization also
improves code generation for AMDGPU: Ganged loads are scheduled into a clause;
stores only await completion of their corresponding load.

The chosen value of 16 performed good in microbenchmarks, values of 8, 32, or
64 would perform similarly, trading off performance in different scenarios.
The testcase updates are autogenerated by utils/update_llc_test_checks.py.

See also:
 - PR introducing this optimization: https://reviews.llvm.org/D46477

Part of SWDEV-455845.

ritter-x2a committed:
fixup! amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

Add a test for combinations of memcpy parameters

ritter-x2a committed:
fixup! fixup! amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

Only include equal alignments in the test.

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-param-combinations.ll:
I've never seen "this diff is too large to display before".

These test functions are all kernels, making the pointer arguments all uniform. It would be somewhat more representative to make these regular functions with VGPR pointer arguments. The cases where known uniform pointers is interesting should then use explicit inreg

ritter-x2a replied:
The test functions are now normal functions instead of kernels and I added another test with `inreg` arguments.
The differences in the generated code between those are not too fundamental, so I skipped most address space combinations and sizes for the sreg test to avoid making it too big.

arsenm replied:
It's only interesting for the SMRD candidate loads from p1/p4 from the sources

arsenm replied:
Which really means p4, without an invariant annotation we don't handle now. You would need to use a kernel for the p1->p4 promotion, so just test the constant address space uniform source case

ritter-x2a replied:
Like this, in `memcpy-sreg.ll`? The previous, more verbose `memcpy-param-combinations-sreg.ll` is removed again.

ritter-x2a committed:
fixup! fixup! fixup! amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

Adjust the test to pass arguments in vregs, add another one that passes arguments explicitly in sregs.

ritter-x2a committed:
fixup! fixup! fixup! fixup! amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

Only test scalar operands for copies/moves from the constant address space 4.

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-sreg.ll:
```diff
@@ -0,0 +1,138 @@
+; Testing codegen for memcpy and memmove with scalar reads.
+
+
+define void @memcpy_p1_p4_sz16_align_4_4(ptr addrspace(1) align 4  %dst, ptr addrspace(4) align 4 readonly inreg %src) {
```
Extra space between align 4 and %st throughout

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-sreg.ll:
```diff
@@ -0,0 +1,138 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+
```
Better name would be scalar-load instead of sreg

arsenm reviewed (approved):
lgtm with nit

ritter-x2a committed:
fixup! fixup! fixup! fixup! fixup! amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

Clean up whitespace and rename memcpy-sreg.ll -> memcpy-scalar-load.ll.

arsenm reviewed (approved):
Only other nit is I would have split the memcpy and memmove cases to separate files

ritter-x2a committed:
fixup! fixup! fixup! fixup! fixup! fixup! amend! [AMDGPU][DAG] Enable ganging up of memcpy loads/stores for AMDGPU

Separate files for memcpy and memmove tests.
